In [135]:
#Importing libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

#Listing the existing files in the dir
for root, dirs, files in os.walk("."):  
    for filename in files:
        print(filename)

1st_assgnmt_collab.ipynb
1st_assignment_intro.ipynb
2nd_assignment_collab.ipynb
2nd_assignment_intro.ipynb
3rd_assignment_evaluation_metrics.ipynb
README.md
COMMIT_EDITMSG
config
description
FETCH_HEAD
HEAD
index
ORIG_HEAD
applypatch-msg.sample
commit-msg.sample
fsmonitor-watchman.sample
post-update.sample
pre-applypatch.sample
pre-commit.sample
pre-push.sample
pre-rebase.sample
pre-receive.sample
prepare-commit-msg.sample
update.sample
exclude
HEAD
master
master
a6f461b1cda82bc2bc64e2f8e752c594cb4b79
5889fd52bdb928eb7b8ef26232706c784a7fb1
8a81d4f821e4a1d51a8c68599d74959a23bb7d
ff03ce761e5a44ccc2b2a825bf574c1bee0631
44966c920b4d862ad58a805a0ece2382922184
05b0aaa2d7ebb03a3e0fb3d246cda98a5b004a
96b16af7d370ec7985b75b69d3af36a913954d
2974eb78edd82b431c3690f7464a503c694356
2041e1411e121aea1a74646ec651e7c9c954db
807046e3874e227078d513b2e90156c2ab1414
fdc207848247f45b640d1b9e61f889fd73a974
9dd7992c843ab1040b900da7e3335ed0ff5593
4eb72f8826545861a52a4848f147bf97669990
2edfd2574cdc6b5ee4d4d59f8

In [106]:
df = pd.read_csv('./files/metrics_ev.csv', skiprows=1).set_index('Unnamed: 0')
#Subsetting predicted and actual ratings
actual = df.iloc[:10,:10].astype(float)
predicted = df.iloc[14:24,:10].astype(float)

In [107]:
#Calculating mean absolute error per user
mae_user = actual.subtract(predicted, axis=1).abs().sum(axis=1).divide(actual.count(axis=1)).sort_values(ascending=False)
mae_user.head(3)
mae_user.tail(3)

Unnamed: 0
442     0.677778
2492    0.588889
2824    0.400000
dtype: float64

Task: Compute the mean absolute error for each user. For example, user 5261 has a mean absolute error of 0.82 (mostly because of a really bad prediction on the Pirates of the Carribean movie). (a) which users have the highest and lowest MAE, and what are those MAE values?

Solution: Users 2267 (0.900000), 3556 (0.842857) and 5261 (0.820000) have the highest MAE while users 442 (0.677778), 2492 (0.588889) and 2824 (0.400000) have the lowest MAE.

In [88]:
#Calculating mean absolute error per movie
mae_movie = actual.subtract(predicted, axis=0).abs().sum(axis=0).divide(actual.count(axis=0)).sort_values(ascending=False)
mae_movie.head(3)
mae_movie.tail(3)

63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)                                         0.460000
11: Star Wars: Episode IV - A New Hope (1977)                                         0.414286
85: Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)    0.380000
dtype: float64

Task: Also compute the mean absolute error associated with the predictions for each movie. For example, Forrest Gump has an MAE of 0.5. (a) which movies have the highest and lowest MAE, and what are those MAE values?

Solution: Movies 12 (1.166667), 22 (1.088889), and 14 (1.020000) have the highest MAE while movies 63 (0.460000), 11 (0.414286) and 85 (0.380000) have the lowest.

In [154]:
#Calculating the overal MAE
nom = actual.subtract(predicted, axis=0).abs().sum().sum()
denom = actual.count().sum()
overall = nom / denom
#Calculating the overal MAE by user
overall_user = mae_user.mean()
#Calculating the overal MAE by movie
overall_movie = mae_movie.mean()

0.7316785714285714

Task: Compute the overall MAE three ways (three decimal places is enough):
(a) the normal MAE, which averages over all predictions
(b) the by-user MAE, which averages the MAEs of each user
(c) the by-movie MAE, which averages the MAEs of each movie

Solution: (a) The overall MAE is 0.7246376811594202.
(b) The overall MAE by user is 0.731107142857143.
(c) The overall MAE by movie is 0.7316785714285714.

In [128]:
#Calculating mean squared errors and root mean square error for each user
mse = actual.subtract(predicted, axis=1).apply(lambda x: x ** 2).sum(axis=1).divide(actual.count(axis=1)).sort_values(ascending=False)
rmse = np.sqrt(mse)
merged = pd.concat([pd.DataFrame(mse), pd.DataFrame(rmse)], axis=1)
merged.columns = ['mse','rmse']
merged

mse      rmse
Unnamed: 0                    
918         1.470000  1.212436
5261        1.170000  1.081665
3556        1.075714  1.037166
2267        0.995000  0.997497
860         0.858571  0.926591
5136        0.791111  0.889444
442         0.641111  0.800694
4942        0.613750  0.783422
2492        0.412222  0.642045
2824        0.224000  0.473286

Task: Next, create a matrix with the squared errors (instead of absolute errors), and compute the MSE and RMSE (root-mean-squared-error) for each user. Excel and other spreadsheets have a built-in function SQRT() to compute square roots. Be sure that you are computing the MSE first, then taking the square root (if you take the square root of each error before averaging it, you end up with MAE again). For example, user 5136 has an MSE of 0.7911 and an RMSE of .8894. (a) which users have the highest and lowest RMSE, and what are those RMSE values?

Solution: User 918 has the highest RMSE (1.212436) while user 2824 has the lowest (0.473286).

In [134]:
#Calculating overall MSE and RMSE.
nom = actual.subtract(predicted, axis=1).apply(lambda x: x ** 2).sum(axis=1).sum()
denom = actual.count(axis=1).sum()
mse_overall = nom / denom
rmse_overall = np.sqrt(mse_overall)
rmse_overall

0.8913432299755174

Task: Then, compute the overall MSE and RMSE across all predictions (just the normal way, no need to compute by-user or by-movie). Again, remember to average all the squared errors to compute MSE, then take the square root of the MSE to get RMSE.

Solution: Overall MSE is 0.7944927536231882 while overall RMSE is 0.8913432299755174.

In [155]:
#Estimating correlation between actual and predicted values
corr = actual.T.corrwith(predicted.T).sort_values(ascending=False); 
corr_overall = corr.mean()
corr

Unnamed: 0
5261    0.920746
442     0.774450
2267    0.633743
918     0.458946
2492    0.413615
4942    0.347857
2824    0.347677
5136    0.169813
3556   -0.095887
860    -0.148181
dtype: float64

Task: 6. Next, compute the correlation between ratings and predictions using the built-in CORREL() function. First, do this for each user (e.g., the correlation between ratings and predictions for user 5136 is 0.1698 -- only about 17%, and therefore not very good). Then, apply the same correlation to the full matrices.
(a) report the top and bottom correlations (the bottom ones may be negative) by user (both user ID and correlation value).
(b) report the overall correlation between ratings and predictions.

Solution: (a) Top correlation is 0.920746 for user 5261 versus bottom correlation of -0.148181 for 860.
(b) The overall correlation between ratings and predictions is 0.382277868831761.